File: orchestrator-node/src/agents/baseAgent.ts
/**
 * Base agent class with common functionality.
 */

import { AgentStatus } from '../models';
import { getLogger, RateLimiter } from '../utils';

export abstract class BaseAgent {
  readonly name: string;
  protected logger: ReturnType<typeof getLogger>;
  protected rateLimiter: RateLimiter;
  private _status: AgentStatus = AgentStatus.PENDING;
  private _lastError?: string;

  constructor(name: string, rateLimiter?: RateLimiter) {
    this.name = name;
    this.logger = getLogger(`agents.${name}`);
    this.rateLimiter = rateLimiter ?? new RateLimiter();
  }

  get status(): AgentStatus {
    return this._status;
  }

  set status(value: AgentStatus) {
    this.logger.info(`Status changed: ${this._status} -> ${value}`);
    this._status = value;
  }

  get lastError(): string | undefined {
    return this._lastError;
  }

  /**
   * Execute the agent's main task.
   * Override this method in subclasses.
   */
  abstract execute(...args: unknown[]): Promise<unknown>;

  /**
   * Run the agent with status tracking and error handling.
   */
  async run<T>(...args: unknown[]): Promise<T> {
    this.status = AgentStatus.RUNNING;
    this._lastError = undefined;

    try {
      const result = await this.execute(...args);
      this.status = AgentStatus.SUCCESS;
      return result as T;
    } catch (error) {
      this._lastError = error instanceof Error ? error.message : String(error);
      this.status = AgentStatus.FAILED;
      this.logger.error(`Agent failed: ${this._lastError}`);
      throw error;
    }
  }

  /**
   * Reset agent state for new execution.
   */
  reset(): void {
    this._status = AgentStatus.PENDING;
    this._lastError = undefined;
  }
}

File: orchestrator/agents/base_agent.py
"""Base agent class with common functionality."""

from abc import ABC, abstractmethod
from typing import Any, Optional

from ..models import AgentStatus
from ..utils import get_logger, RateLimiter


class BaseAgent(ABC):
    """
    Abstract base class for all agents.
    
    Provides common functionality like logging, rate limiting,
    and status management.
    """
    
    def __init__(self, name: str, rate_limiter: Optional[RateLimiter] = None):
        self.name = name
        self.logger = get_logger(f"agents.{name}")
        self.rate_limiter = rate_limiter or RateLimiter()
        self._status = AgentStatus.PENDING
        self._last_error: Optional[str] = None
    
    @property
    def status(self) -> AgentStatus:
        """Current agent status."""
        return self._status
    
    @status.setter
    def status(self, value: AgentStatus):
        self.logger.info(f"Status changed: {self._status} -> {value}")
        self._status = value
    
    @property
    def last_error(self) -> Optional[str]:
        """Last error message if agent failed."""
        return self._last_error
    
    @abstractmethod
    async def execute(self, *args, **kwargs) -> Any:
        """
        Execute the agent's main task.
        
        Override this method in subclasses to implement
        agent-specific logic.
        """
        pass
    
    async def run(self, *args, **kwargs) -> Any:
        """
        Run the agent with status tracking and error handling.
        
        This is the main entry point that wraps execute()
        with proper status management.
        """
        self.status = AgentStatus.RUNNING
        self._last_error = None
        
        try:
            result = await self.execute(*args, **kwargs)
            self.status = AgentStatus.SUCCESS
            return result
        except Exception as e:
            self._last_error = str(e)
            self.status = AgentStatus.FAILED
            self.logger.error(f"Agent failed: {e}")
            raise
    
    def reset(self):
        """Reset agent state for new execution."""
        self._status = AgentStatus.PENDING
        self._last_error = None
